src/components/Sketches/components/EditSketchModal.js

```diff
@@ -0,0 +1,323 @@
+import * as fetch from "../../../lib/fetch.js";
+
+class EditSketchModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      newLanguage: -1,
+      newName: -1,
+      newThumbnail: -1,
+      spinner: false,
+      error: "",
+      onThumbnails: false,
+    };
+  }
+  componentWillMount() {}
+  componentDidUpdate() {}
+
+  closeModal = () => {
+    if (this.props.onClose && {}.toString.call(this.props.onClose) === "[object Function]") {
+      this.props.onClose();
+    }
+    this.setState({
+      next: false,
+      newLanguage: -1,
+      newName: -1,
+      newThumbnail: -1,
+      error: "",
+      onThumbnails: false,
+      spinner: false,
+    });
+  };
+
+  // Next two bad____input are copied (and slightly modified) from create sketch modal
+
+  badNameInput = () => {
+    if (this.state.newName === -1) {
+      return false;
+    }
+    if (!this.state.newName) {
+      this.setState({ error: "Name is required" });
+      return true;
+    }
+    if (this.state.newName.length > 15) {
+      this.setState({ error: "Name must be 15 characters or less" });
+      return true;
+    }
+    return false;
+  };
+
+  badLanguageInput = () => {
+    if (this.state.newLanguage === -1) {
+      return false;
+    }
+    if (!this.state.newLanguage) {
+      this.setState({ error: "Please select a language" });
+      return true;
+    }
+
+    let notFound = true;
+    for (var i = 0; i < LanguageDropdownValues.length; i++) {
+      if (
+        this.state.newLanguage.display === LanguageDropdownValues[i].display &&
+        this.state.newLanguage.value === LanguageDropdownValues[i].value
+      ) {
+        notFound = false;
+        break;
+      }
+    }
+    if (notFound) {
+      this.setState({ error: "Invalid language selected" });
+      return true;
+    }
+
+    return false;
+  };
+
+  handleSubmitEdit = async e => {
+    e.preventDefault();
+
+    if (this.badNameInput() || this.badLanguageInput()) {
+      // note it's impossible to have a bad thumbnail input
+      return;
+    }
+
+    let data = {};
+    let doUpdate = false;
+
+    if (this.state.newLanguage !== -1) {
+      data["language"] = this.state.newLanguage.value;
+      doUpdate = true;
+    }
+    if (this.state.newName !== -1) {
+      data["name"] = this.state.newName;
+      doUpdate = true;
+    }
+    if (this.state.newThumbnail !== -1) {
+      data["thumbnail"] = this.state.newThumbnail;
+      doUpdate = true;
+    }
+    if (doUpdate) {
+      let updateData = {};
+      updateData[this.props.sketchKey] = data;
+      try {
+        fetch
+          .updatePrograms(this.props.uid, updateData)
+          .then(res => {
+            return res.json();
+          })
+          .then(json => {
+            if (!json.ok) {
+              this.setState({
+                spinner: false,
+                error: json.error || "Failed to create sketch, please try again later",
```

Comment:
Failed to **edit** sketch, please try again later

Reply:
Fixed!